pkg/openstack/routes.go

```diff
@@ -35,97 +36,127 @@ import (
 		return nil, errors.ErrNoRouterID
 	}
 
 	return &Routes{
-		compute:        compute,
-		network:        network,
-		opts:           opts,
-		networkingOpts: networkingOpts,
+		network: network,
+		os:      os,
 	}, nil
 }
 
 // ListRoutes lists all managed routes that belong to the specified clusterName
 func (r *Routes) ListRoutes(ctx context.Context, clusterName string) ([]*cloudprovider.Route, error) {
 	klog.V(4).Infof("ListRoutes(%v)", clusterName)
 
-	nodeNamesByAddr := make(map[string]types.NodeName)
-	err := foreachServer(r.compute, servers.ListOpts{}, func(srv *servers.Server) (bool, error) {
-		interfaces, err := getAttachedInterfacesByID(r.compute, srv.ID)
-		if err != nil {
-			return false, err
-		}
-
-		addrs, err := nodeAddresses(srv, interfaces, r.networkingOpts)
-		if err != nil {
-			return false, err
-		}
-
-		name := mapServerToNodeName(srv)
-		for _, addr := range addrs {
-			nodeNamesByAddr[addr.Address] = name
-		}
+	if r.os.nodeInformerHasSynced == nil || !r.os.nodeInformerHasSynced() {
+		return nil, errors.ErrNoNodeInformer
+	}
 
-		return true, nil
-	})
+	nodes, err := r.os.nodeInformer.Lister().List(labels.Everything())
 	if err != nil {
 		return nil, err
 	}
 
 	mc := metrics.NewMetricContext("router", "get")
-	router, err := routers.Get(r.network, r.opts.RouterID).Extract()
+	router, err := routers.Get(r.network, r.os.routeOpts.RouterID).Extract()
 	if mc.ObserveRequest(err) != nil {
 		return nil, err
 	}
 
 	var routes []*cloudprovider.Route
 	for _, item := range router.Routes {
-		nodeName, foundNode := nodeNamesByAddr[item.NextHop]
-		if !foundNode {
-			nodeName = types.NodeName(item.NextHop)
-		}
+		nodeName, foundNode := getNodeNameByAddr(item.NextHop, nodes)
 		route := cloudprovider.Route{
 			Name:            item.DestinationCIDR,
-			TargetNode:      nodeName, //contains the nexthop address if node was not found
+			TargetNode:      nodeName, //contains the nexthop address if node name was not found
 			Blackhole:       !foundNode,
 			DestinationCIDR: item.DestinationCIDR,
 		}
 		routes = append(routes, &route)
 	}
 
+	// detect router's private network ID for further VM ports filtering
+	r.Lock()
+	r.networkIDs, err = getRouterNetworkIDs(r.network, r.os.routeOpts.RouterID)
+	r.Unlock()
+	if err != nil {
+		return nil, err
+	}
+
 	return routes, nil
 }
 
-func foreachServer(client *gophercloud.ServiceClient, opts servers.ListOptsBuilder, handler func(*servers.Server) (bool, error)) error {
-	mc := metrics.NewMetricContext("server", "list")
-	pager := servers.List(client, opts)
+func getRouterNetworkIDs(network *gophercloud.ServiceClient, routerID string) ([]string, error) {
+	opts := ports.ListOpts{
+		DeviceID: routerID,
+	}
+	mc := metrics.NewMetricContext("port", "list")
+	pages, err := ports.List(network, opts).AllPages()
+	if mc.ObserveRequest(err) != nil {
+		return nil, err
+	}
+	ports, err := ports.ExtractPorts(pages)
+	if err != nil {
+		return nil, err
+	}
+
+	var networkIDs []string
```

Comment:
this does not contain unique values for networkid?

Reply:
duplicate values are less probable, since router cannot handle the same network more than once.